stdlib/domain.ml

```diff
@@ -112,3 +112,25 @@ let get_id { domain; _ } = domain
 
 let self () = Raw.self ()
 
+module TLS = struct
+
+  type 'a key = int ref
+
+  external get_tls_list : unit -> (int ref * Obj.t) list
+    = "caml_domain_tls_get"
+
+  external set_tls_list : (int ref * Obj.t) list  -> unit
+    = "caml_domain_tls_set"
+
+  let new_key () = ref 0
+
+  let set k x =
+    let cs = Obj.repr x in
+    let old = get_tls_list () in
+    set_tls_list @@ (k,cs)::(List.remove_assq k old)
```

Comment:
I'm not particularly fond of this. In the worst case, If you remove the last element in the list, then the whole spine of the list will need to be allocated again. I still recommend making the second element of the pair a ref.